File: python/shopping_cart.py
from .model_objects import ProductQuantity
from .discount_service import DiscountService


class ShoppingCart:
    """Represents a shopping cart containing items to be purchased"""

    def __init__(self):
        self._items = []
        self._product_quantities = {}
        self._discount_service = DiscountService()

    @property
    def items(self):
        """Get list of all items in the cart"""
        return self._items

    @property
    def product_quantities(self):
        """Get dictionary of products and their total quantities"""
        return self._product_quantities

    def add_item(self, product):
        """Add a single item to the cart"""
        self.add_item_quantity(product, 1.0)

    def add_item_quantity(self, product, quantity):
        """Add a product with specified quantity to the cart"""
        if quantity < 0:
            raise ValueError("Quantity cannot be negative")

        self._items.append(ProductQuantity(product, quantity))

        if product in self._product_quantities:
            self._product_quantities[product] += quantity
        else:
            self._product_quantities[product] = quantity

    def apply_discounts(self, receipt, offers, catalog):
        """Apply all applicable discounts to the receipt"""
        self._discount_service.apply_discounts(
            receipt,
            self._product_quantities,
            offers,
            catalog
        )

File: src/SupermarketReceipt/ShoppingCart.cs
using System.Collections.Generic;

namespace SupermarketReceipt;

public class ShoppingCart
{
    private readonly List<ProductQuantity> _items = new List<ProductQuantity>();
    private readonly Dictionary<Product, Quantity> _productQuantities = new Dictionary<Product, Quantity>();


    public List<ProductQuantity> GetItems()
    {
        return new List<ProductQuantity>(_items);
    }

    public Dictionary<Product, Quantity> GetProductQuantities()
    {
        return new Dictionary<Product, Quantity>(_productQuantities);
    }

    public void AddItem(Product product)
    {
        AddItemQuantity(product, 1.0);
    }


    public void AddItemQuantity(Product product, double quantity)
    {
        var productQuantity = new Quantity(quantity, product.Unit);
        _items.Add(new ProductQuantity(product, productQuantity));
        if (_productQuantities.ContainsKey(product))
        {
            var currentQuantity = _productQuantities[product];
            var newQuantity = currentQuantity with { Amount = currentQuantity.Amount + productQuantity.Amount };
            _productQuantities[product] = newQuantity;
        }
        else
        {
            _productQuantities.Add(product, productQuantity);
        }
    }

}